### PR TITLE
fix: Browser dialog prompts in app

### PR DESCRIPTION
**Pull request by Daniel-Cross**
## **Description**

Added a function that checks to see if the Webview Tab is currently in focus. If the user navigates away from the webview at any point, the app will stop any JS execution in the webview. If the user remains in the webview tab, JS execution will continue as expected.

## **Related issues**

Fixes: [Browser Dialog prompts can overlay outside of tab](https://github.com/orgs/MetaMask/projects/60/views/16?pane=issue&itemId=43794381)

## **Manual testing steps**

1. Download the following file and visit it in the MM Mobile browser: https://github.com/MetaMask/temp-poc-templates-repo/blob/obs3rv3r-pocs/metamask-poc.html
2. Upon loading the page, switch to another part of the app, such as the settings page or the home page.
3. See that the browser UI prompts still show up over the application
4. Do steps 1-2 again but this time visit another tab in the browser.
5. See that the prompts show up over the other tab

### **After**

https://github.com/user-attachments/assets/95f7aee7-1e70-4a1f-8c43-55a95d3dccc8

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/components/Views/BrowserTab/index.js [hunk 1]
```diff
@@ -706,6 +708,21 @@ export const BrowserTab = (props) => {
     };
   }, [goBack, isTabActive, props.navigation]);
 
+  useEffect(() => {
+    if (webviewRef.current) {
```
Comment (sethkfman) [inline app/components/Views/BrowserTab/index.js:712]: Could you break this out into a utils function and add a unit test? 

It is pretty simple but will help document this as an expected behavior.

Comment (Daniel-Cross) [reply app/components/Views/BrowserTab/index.js]: Moved the function into it's own file and added some unit tests.

**Comment by tommasini** [comment]
Hey Dan! Can we have recordings of the in app browser working? 
I tried a solution like this on the past: https://github.com/MetaMask/metamask-mobile/pull/7809, I believe this can bring problems, the thread were we discussed this is no longer visible but we should test this solution by communicating with known dapps, using the SDK, wallet connect & deeplinks

**Comment by tommasini** [comment]
Found the issue that I had: https://github.com/MetaMask/metamask-mobile/pull/7893

Test case:
 "when the user is on in app browser and needs to switch network, it will disable the javascript and that will make the dapp be misaligned."

**Review by sethkfman — DISMISSED** [review]
LGTM

File: ios/Podfile.lock [hunk 2]
```diff
@@ -540,8 +540,8 @@ PODS:
     - Base64
     - GZIP
     - React-Core
-  - react-native-in-app-review (3.2.3):
-    - React
+  - react-native-in-app-review (4.3.3):
```
Comment (tommasini) [inline ios/Podfile.lock:543]: I think this podflie lock change was commited by mistake, or it was intentional?

File: app/components/Views/BrowserTab/index.js [hunk 3]
```diff
@@ -706,6 +710,17 @@ export const BrowserTab = (props) => {
...
 
+  useEffect(() => {
+    handleWebViewFocus({
+      webviewRef,
+      isFocused,
+      chainId: props.chainId,
+      previousChainId,
+    });
+    setPreviousChainId(props.chainId);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (tommasini) [inline app/components/Views/BrowserTab/index.js:721]: Can we avoid disabling this rule? This rule helps us performance wise, and it's something we want to focus on!

Maybe creating an useCallback and split the logic can help

File: ios/Podfile.lock (hunk 2 above)
Comment (Daniel-Cross) [reply ios/Podfile.lock:543]: This is most likely unintentional. I'll revert. Thanks

File: app/util/browser/webViewFocus.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+
+interface WebViewFocus {
+  webviewRef: React.RefObject<{
+    injectJavaScript: (script: string) => void;
+    stopLoading: () => void;
+    reload: () => void;
+  }>;
+  isFocused: boolean;
+  chainId: string;
```
Comment (tommasini) [inline app/util/browser/webViewFocus.ts:10]: For consistency let's type chainId and previousChainId as `Hex` from metamask/utils

File: app/util/browser/webViewFocus.test.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+  it('should reload the webview if chainId has changed', () => {
+    const stopLoading = jest.fn();
+    const reload = jest.fn();
+    const webviewRef = { current: { stopLoading, reload } };
+
+    handleWebViewFocus({
+      webviewRef,
+      isFocused: true,
+      chainId: '2',
```
Comment (tommasini) [inline app/util/browser/webViewFocus.test.ts:23]: Let's use hexadecimal values for chain id for consistency

File: app/util/browser/webViewFocus.test.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  it('should not reload the webview if chainId has not changed', () => {
+    const stopLoading = jest.fn();
+    const reload = jest.fn();
+    const webviewRef = { current: { stopLoading, reload } };
+
+    handleWebViewFocus({
+      webviewRef,
+      isFocused: true,
+      chainId: '1',
```
Comment (tommasini) [inline app/util/browser/webViewFocus.test.ts:39]: Here too

File: app/util/browser/webViewFocus.test.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+  it('should do nothing if webviewRef.current is null', () => {
+    const stopLoading = jest.fn();
+    const reload = jest.fn();
+    const webviewRef = { current: null };
+
+    handleWebViewFocus({
+      webviewRef,
+      isFocused: true,
+      chainId: '1',
```
Comment (tommasini) [inline app/util/browser/webViewFocus.test.ts:55]: Here too

**Review by tommasini — DISMISSED** [review]
LGTM! After addressing the chain id for hexadecimal values, this looks good to me!

**Review by tommasini — APPROVED** [review]

**Pull request merged**